src/Remote.cc

```diff
@@ -0,0 +1,81 @@
+/**
+ * \file   Remote.cc
+ * \author Lars Fröhlich
+ * \date   Created on January 15, 2024
+ * \brief  Implementation of the Remote class.
+ *
+ * \copyright Copyright 2024 Deutsches Elektronen-Synchrotron (DESY), Hamburg
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 2.1 of the license, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include <git2.h>
+#include <gul14/cat.h>
+#include <gul14/string_util.h>
+
+#include "libgit4cpp/Error.h"
+#include "libgit4cpp/Remote.h"
+#include "libgit4cpp/wrapper_functions.h"
+#include "credentials_callback.h"
+
+using gul14::cat;
+
+namespace git {
+
+Remote::Remote(LibGitRemote&& remote_ptr)
+    : remote_{ std::move(remote_ptr) }
+{
+    if (remote_ == nullptr)
+        throw Error{ "Remote pointer may not be null" };
+
+    name_ = gul14::safe_string(git_remote_name(remote_.get()), 512);
+    url_ = gul14::safe_string(git_remote_url(remote_.get()), 512);
```

Comment:
> What is the magic of `512`? And why is it not `500` (I will never understand why limits should be power-of-2 for no technical reason)?

No magic. Whatever.

> I also do not understand why you expect an un-terminated string here?

I do not, but I do expect nullptr: "a pointer to the name or NULL for in-memory remotes"

> `libgit` does make sure it is valid (well, at least it also calls `strlen()` on it, unlimited, before you "do the same" but more anxious. Also other returned strings are not handed like this (e.g. from `git_remote_list()`)?

It is not necessarily valid. It might be null.

Reply:
But `safe_string` also also about a length limit (presumably to avoid ... what? Non-terminated string crashes?)

Then why not the much clearer
```diff
-name_ = gul14::safe_string(git_remote_name(remote_.get()), 512);
+auto n = git_remote_name(remote_.get());
+name_ = n ? n : ""; // or something? Is "" a good name for file remotes?
```
that does not break for otherwise working idiotic long names :->

Reply:
> Then why not the much clearer
> 
> ```diff
> -name_ = gul14::safe_string(git_remote_name(remote_.get()), 512);
> +auto n = git_remote_name(remote_.get());
> +name_ = n ? n : ""; // or something? Is "" a good name for file remotes?
> ```
> 
> that does not break for otherwise working idiotic long names :->

Fair enough, changed.